cirq-core/cirq/sim/simulator_base_test.py

```diff
@@ -192,3 +256,81 @@ def test_run_non_terminal_measurement():
     sim = CountingSimulator()
     r = sim.run(cirq.Circuit(cirq.X(q0), cirq.measure(q0), cirq.X(q0)), repetitions=2)
     assert np.allclose(r.measurements['0'], [[1], [1]])
+
+
+def test_integer_initial_state_is_split():
+    sim = SplittableCountingSimulator()
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert len(set(args.values())) == 3
+    assert args[q0].state == 1
+    assert args[q1].state == 0
+    assert args[None].state == 0
+
+
+def test_integer_initial_state_is_not_split_if_disabled():
+    sim = SplittableCountingSimulator(split_untangled_states=False)
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert isinstance(args, SplittableCountingActOnArgs)
+    assert args.state == 2
+
+
+def test_integer_initial_state_is_not_split_if_impossible():
+    sim = CountingSimulator()
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert isinstance(args, CountingActOnArgs)
+    assert not isinstance(args, SplittableCountingActOnArgs)
+    assert args.state == 2
+
+
+def test_non_integer_initial_state_is_not_split():
+    sim = SplittableCountingSimulator()
+    args = sim._create_act_on_args('state', (q0, q1))
+    assert len(set(args.values())) == 2
+    assert args[q0].state == 'state'
+    assert args[q1] is args[q0]
+    assert args[None].state == 0
+
+
+def test_entanglement_causes_join():
+    sim = SplittableCountingSimulator()
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert len(set(args.values())) == 3
+    args.apply_operation(cirq.CNOT(q0, q1))
+    assert len(set(args.values())) == 2
+    assert args[q0] is args[q1]
+    assert args[None] is not args[q0]
+
+
+def test_measurement_causes_split():
+    sim = SplittableCountingSimulator()
+    args = sim._create_act_on_args('state', (q0, q1))
+    assert len(set(args.values())) == 2
+    args.apply_operation(cirq.measure(q0))
+    assert len(set(args.values())) == 3
+    assert args[q0] is not args[q1]
+    assert args[q0] is not args[None]
+
+
+def test_measurement_does_not_split_if_disabled():
+    sim = SplittableCountingSimulator(split_untangled_states=False)
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert isinstance(args, SplittableCountingActOnArgs)
+    args.apply_operation(cirq.measure(q0))
+    assert isinstance(args, SplittableCountingActOnArgs)
+
+
+def test_measurement_does_not_split_if_impossible():
+    sim = CountingSimulator()
+    args = sim._create_act_on_args(2, (q0, q1))
+    assert isinstance(args, CountingActOnArgs)
+    assert not isinstance(args, SplittableCountingActOnArgs)
+    args.apply_operation(cirq.measure(q0))
+    assert isinstance(args, CountingActOnArgs)
+    assert not isinstance(args, SplittableCountingActOnArgs)
```

Comment:
See the comment on line 274: having `__getitem__` for `ActOnArgs` would allow us to more directly test that `args[q0] == args[q1]` in these tests.

To confirm: is there any case where a simulator which uses `ActOnArgs` would switch to using `ActOnArgsContainer`, or vice versa?

Reply:
> is there any case where a simulator which uses ActOnArgs would switch to using ActOnArgsContainer, or vice versa

No, the OperationTarget is a single instance throughout the simulation.